design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,163 @@
+
+### Reading virtual chunks
+
+* When a user create a repository instance, they can specify credentials for each `VirtualChunkContainer`, pointing to them by name.
+* They can also set default `VirtualChunkContainer` credentials, that will be used for any container not specifically set
+* If no credentials are specified the same as for the main `Storage` will be used (if platforms are the same)
+* On read of a virtual chunk:
+  * The `VirtualChunkContainer` is resolved from configuration.
+  * An object storage client is instantiated. We'll need a cache here, so we don't keep creating new instances for each chunk
+  * The client is initialized with the credentials passed (or defaulted).
+  * The container template is expanded with the `template_args` in the virtual chunks. If there are no enough template arguments or some are None, they are replaced by the default template arguments in the container. If there are too many, last ones are ignored.
+  * The chunk is fetched from the results URL.
+
+### Writing virtual chunks
+
+* Users must first update the repo configuration to include the needed `VirtualChunkContainers`.
+* When the user writes a virtual chunk ref, they pass in the `VirtualChunkRef` the index for the container and the template expansion arguments. An error is produced if the container is not found.
+* This is a bit inconvenient, having to declare the container before being able to write. An alternative would be to automatically create a container that includes only the bucket name as template. On writes we try to match all templates against the url, and use the first one that matches, if non does, we create a new container. We could add that as a new function in the future if it's not required on first version.
+
+## Check requirements are satisfied
+
+1. > Virtual chunks can be stored in all supported object stores
+
+   Now the virtual chunk containers inform what is the platform, and we can create the appropriate object store client at read time. These clients are cached for performance.
+2. > Multiple credentials sets for virtual chunk resolution
+
+    Now the user initializes a Repo instance by passing, if needed, a credential set for each virtual chunk container. If a chunk is read for which there are no credentials, we default to a common set of credentials for virtual chunks, and ultimately, to the Storage credentials if on the same platform.
+3. > Optimized manifests where the same prefix doesn't need to be repeated for every virtual chunk
+  
+    For large prefixes, the template + arguments creates some savings. For example, a template for virtual zarr like `some-long-bucket-name/with-some-even-longer-prefix/and-more-nesting/some-long-array-name-for-the-dataset/c/{}/{}/{}`.
```

Comment:
Again, I need a lot of convincing on this in the presence of decent compression, say Zstd.

```python
strings = "".join([f"some-long-bucket-name/with-some-even-longer-prefix/and-more-nesting/some-long-array-name-for-the-dataset/c/{i}/{i}/{i}" for i in range(10000)])
out = cramjam.zstd.compress(strings.encode())
len(strings) # 1206670
len(out)   # 19052, 1.9 bytes per entry
```

Reply:
Remember this was an optimization for the in-memory datastructure. Coupling it with the on-disk format was a mistake (as a consequence of Icechunk still not having a separation layer there). But even so, this is not the right way to optimize this. I dropped the requirement, we can absolutely do this optimization later. Thanks for the feedback on this, it changed my mind.